### ISSUE TITLE
Freeze when resizing AppTabs during Close Intercept on develop (working on 2.3.4)

### ISSUE THREAD

**Issue by xypwn** [fyne-io/fyne#3922]
### Checklist

- [X] I have searched the [issue tracker](https://github.com/fyne-io/fyne/issues) for open issues that relate to the same problem, before opening a new one.
- [X] This issue only relates to a single bug. I will open new issues for any other problems.

### Describe the bug

In my app, I want to wait for the current task to finish cleanly, waiting to close until the waitgroup is done.
When I close the window and don't touch it until the task exits, everything works. But if I resize the window while still waiting for the task to finish, it irreversibly freezes up.

### How to reproduce

- Run the example code
- Close the window (in this case within the first 10s)
- Before it actually closes, try resizing (see video attachment)

Voi-là, you just froze your app

### Screenshots

[recording_20230527-200714.webm](https://github.com/fyne-io/fyne/assets/54681180/9d8864b0-3541-4928-8733-0f55f5a2fc4d)

### Example code

```go
package main

import (
    "context"
    "fmt"
    "sync"
    "time"

    "fyne.io/fyne/v2"
    "fyne.io/fyne/v2/app"
    "fyne.io/fyne/v2/container"
    "fyne.io/fyne/v2/theme"
    "fyne.io/fyne/v2/widget"
)

func main() {
    app := app.New()
    window := app.NewWindow("Window Title")

    ctx, cancel := context.WithCancel(context.Background())
    var wg sync.WaitGroup

    wg.Add(1)
    go func() {
        defer wg.Done()
        fmt.Println("TASK: Starting to do something")
        for i := 0; i < 10; i++ {
            time.Sleep(1 * time.Second)
            select {
            case <-ctx.Done():
                err := ctx.Err()
                if err != nil {
                    fmt.Println("TASK: Error:", err)
                    return
                }
                fmt.Println("TASK: Interrupted")
            default:
            }
        }
        fmt.Println("TASK: Done")
    }()

    appTabs := container.NewAppTabs(
        container.NewTabItemWithIcon(
            "Downloads",
            theme.DownloadIcon(),
            widget.NewButton("Downloads", func() {}),
        ),
        container.NewTabItemWithIcon(
            "Preferences",
            theme.SettingsIcon(),
            widget.NewButton("Preferences", func() {}),
        ),
    )
    appTabs.SetTabLocation(container.TabLocationLeading)

    window.SetContent(appTabs)

    window.Resize(fyne.NewSize(800, 450))

    window.SetCloseIntercept(func() {
        go func() {
            cancel()
            fmt.Println("MAIN: Window closed, waiting for task to finish...")
            wg.Wait()
            fmt.Println("MAIN: Finished, exiting")
            window.Close()
        }()
    })

    window.ShowAndRun()
}
```

### Fyne version

v2.3.4-rc1.0.20230526221837-732ffb69a21d

### Go compiler version

go1.20.3

### Operating system and version

Manjaro Linux / Wayland

### Additional Information

Strangely, when I remove the AppTabs and replace them with a label, this issue doesn't occur.

**Comment by andydotxyz** [comment]
Retitled as "SetCloseIncercept broken" was rather misleading.

**Comment by toaster** [comment]
On MacOS, it isn’t even necessary to resize the window. The close attempt within the first ten seconds freezes the app.  After a while, the frozen app crashes with a stack overflow. Closing after more then ten seconds works fine.

**Comment by toaster** [comment]
The MacOS error does not happen when using an error group (`golang.org/x/sync/errgroup`). Since this is simply a warpper around the plain wait group, there seems to be something odd happening.

**Comment by toaster** [comment]
Okay, seems I was too slow. When increasing the time between the selects, I can reproduce the error with the error group in place but I actually have to attempt a resize.

**Comment by Jacalz** [comment]
For anyone looking at this, you might have some luck doing the following to get a stack trace once the application freezes up:

1. Found the pid for the program using `ps -a | grep program` and substituting "program" with your program name.

2. Run `kill -QUIT pid` (this sends a SIGQUIT instead of outright killing it) while substituting "pid" for the pid you found above. This should get you a stack trace for where the application is stuck.

(This solution does not work on Windows as far as I know but you might be able to do something similar using the task manger)

**Comment by andydotxyz** [comment]
It looks like it creates an infinite loop trying to destroy widgets that were already dead...

What would folk feel about a fix along these lines?

```go
diff --git a/internal/driver/util.go b/internal/driver/util.go
index b010d5934..c8b43ceff 100644
--- a/internal/driver/util.go
+++ b/internal/driver/util.go
@@ -157,7 +157,9 @@ func walkObjectTree(
        case *fyne.Container:
                children = co.Objects
        case fyne.Widget:
-               children = cache.Renderer(co).Objects()
+               if cache.IsRendered(co) {
+                       children = cache.Renderer(co).Objects()
+               }
        }
 
        if _, ok := obj.(fyne.Scrollable); ok {
```

**Comment by andydotxyz** [comment]
Fixed in v2.4.0-rc1
